### PR TITLE
feat(cardano-sevices): allow api compatible version requests

### PR DESCRIPTION
**Pull request by mirceahasegan**
# Context

Given /v[major].[minor].[patch],
Server accepts requests to urls that have <= minor version number
Server does not accept requests to urls that have different major version number
Patch version is ignored

### CODE REVIEW

File: packages/cardano-services/src/Http/HttpServer.ts [hunk 1]
```diff
@@ -92,9 +109,10 @@ export class HttpServer extends RunnableModule {
     for (const service of this.#dependencies.services) {
       await service.initialize();
       const serviceVersionPath = service.apiVersionPath();
-      const serviceBaseUrl = path.join(serviceVersionPath, service.slug);
-      this.app.use(serviceBaseUrl, service.router);
-      this.logger.info(`Using ${serviceBaseUrl}`);
+      const majorVersion = serviceVersionPath.split('.')[0];
+      const serviceBaseUrlRegex = new RegExp(`^${majorVersion}\\.\\d+\\.\\d+\\/${service.slug}`);
+      this.app.use(serviceBaseUrlRegex, this.checkReqVersion(serviceVersionPath), service.router);
+      this.logger.info(`Using ${serviceBaseUrlRegex}`);
```
Comment (mirceahasegan) [inline packages/cardano-services/src/Http/HttpServer.ts:115]: I am not sure what we want to display here. Maybe we should keep displaying the exact version. What do you think?

File: packages/cardano-services/src/Http/HttpServer.ts [hunk 2]
```diff
@@ -61,6 +61,23 @@ export class HttpServer extends RunnableModule {
     return result;
   }
 
+  private checkReqVersion(version: string) {
+    return (req: express.Request, res: express.Response, next: express.NextFunction) => {
+      const [major, minor] = version.split('.');
+      const versionRegex = new RegExp(`^${major}\\.(\\d+)\\.(\\d+)`);
```
Comment (mkazlauskas) [inline packages/cardano-services/src/Http/HttpServer.ts:67]: nit: those could be a cached in outer scope. Also no need to capture 'patch' in a group

File: packages/cardano-services/src/Http/HttpServer.ts (hunk 1 above)
Comment (mkazlauskas) [reply packages/cardano-services/src/Http/HttpServer.ts:115]: That looks fine to me, assuming it's printing the regex in a readable way

Comment (mirceahasegan) [reply packages/cardano-services/src/Http/HttpServer.ts:115]: It looks like: `Using /^\/v15\.\d+\.\d+\/utxo/`

Comment (mkazlauskas) [reply packages/cardano-services/src/Http/HttpServer.ts:115]: That's fine by me :)

File: packages/cardano-services/src/Http/HttpServer.ts [hunk 3]
```diff
@@ -61,6 +61,23 @@ export class HttpServer extends RunnableModule {
...
+      const match = req.baseUrl.match(versionRegex);
+
+      if (match) {
+        const [requestMinor] = match.slice(1).map(Number);
+        if (requestMinor <= Number(minor)) {
+          return next();
+        }
+      }
+
+      return res.status(400).send('Invalid version');
```
[the remark this answers is not in the document]
Comment (mirceahasegan) [reply packages/cardano-services/src/Http/HttpServer.ts]: done here: [c7bb59e](https://github.com/input-output-hk/cardano-js-sdk/pull/1022/commits/c7bb59ec2741ee019543e7aaffbea3e18bda4008)

File: packages/cardano-services/src/Http/HttpServer.ts (hunk 2 above)
Comment (mirceahasegan) [reply packages/cardano-services/src/Http/HttpServer.ts]: done here: [c7bb59e](https://github.com/input-output-hk/cardano-js-sdk/pull/1022/commits/c7bb59ec2741ee019543e7aaffbea3e18bda4008)

**Review by mchappell — APPROVED** [review]

**Review by mkazlauskas — DISMISSED** [review]
Nice work! :rocket:

**Review by mkazlauskas — APPROVED** [review]

**Review by mchappell — APPROVED** [review]

**Pull request merged**